### ISSUE TITLE
Nesting configuration: The default values cannot be seen by lowest level tasks

### ISSUE THREAD

**Issue by Aiky30** [pyinvoke/invoke#470]
When trying the following example: http://docs.pyinvoke.org/en/latest/concepts/configuration.html#configuration-overriding, the lowest level can indeed override defaults set in higher order, however i believe that there is an issue because the defaults set in the higher order cannot be seen in the lower levels. 

If you were to inspect the context in the lowest level with nothing overridden it will be only the defaults of invoke and not defaults set by the higher order modules. 

Is this by design?

I believed that the configs are all compiled and will be available to all scripts that access the context, including any defaults set higher up. This would also be the reasons behind having to compile all of the Collections into one int he lowest level.

I am happy to assist with this further if you need me to create an example showing the issue. Please tell me the method in which this is normally done for you.

**Comment by bitprophet** [comment]
An example would be useful, yes! I'm not 100% sure I get what you're talking about. It _sounds like_ your issue is the fact that collection-level config is stored in the config object as a single, squashed/flattened dict made up out of all levels in the task namespace. (Either that this is making debugging harder for you, or is actively causing a problem with the final resulting config data.)

FWIW that design is 1st-draft quality and the rest of the Config design has evolved to be a lot cleaner, so I'd like (sometime...) to fix up the collection level too (so it's presented as distinct layers and one can inspect exactly which bits came from which collection objects).

Again, though, please provide a concrete example of what you're trying to do & how it's not working. Thanks!

**Comment by Aiky30** [comment]
Thanks for the quick reply regarding this. 

I may not have been very clear with my wording. Hopefully the following example will demonstrate the issues that I'm facing. 

It may be better to first explain that I wish to create a tool that has a default configuration that can be overridden or left as default. The tool / app can be used on various machines each with a slightly different minimal config.

As per your request: https://github.com/Aiky30/invoke-nested-config-example

I have provided a stripped down example that contains two configs. One config is within base/base_tasks.py. This is where I see defaults going. I then have in the outer level a standard invoke tasks.py, containing an outer task which I expected to print the defaults in base as well as the config set in invoke.py. Instead It cannot see the defaults set in base. I have purposely not overridden the base config in the outer config to show that the base config is not available to the outer level. Is this correct behaviour or an issue?

My understanding is that the config is compiled and should be the same for all tasks no matter where they are in the hierarchy. I'm unsure if my understanding is correct or whether the methods that I'm trying to use are the most up to date way with your plans.

**Comment by bitprophet** [comment]
Gotcha, thanks! Yes, your understanding is off from current reality - config tasks _are not_ all the same, they reflect their position in the collection/namespace tree. The docs try to make this clear: http://docs.pyinvoke.org/en/latest/concepts/configuration.html#collection-based-configuration

The decision was made that configs closer to the root namespace win over configs closer to the task (the 'leaf' of the tree.) This is to support things like distributing an 'inner' collection which defines its own default values, but where the including 'outer' collection can override them if needed - much like calling a function that has default keyword argument values.

You seem to expect the opposite - that the outer layers of the namespace tree 'inherit' data defined in the inner layers. Unfortunately that would make distributing reusable sub-namespaces rather difficult and probably error-prone.

----

At the moment, the intention is for you to distribute your tool as a "packaged" collection or namespace, and for consumers of the tool to include it into their own `tasks.py`'s namespace; then they have the option of overriding your distributed defaults (again in that root namespace object) or not.

A real-world example of this is how my projects use `invocations.docs`: https://github.com/pyinvoke/invocations/blob/9e1cae2420c07adc26a0d8dc5b555cbebac7a2f5/invocations/docs.py

- It defines a bunch of tasks (the top half of the file) expecting that `c.sphinx.target` and similar variables are set in the config
- It defines a local default namespace, with collection-level config assuming your docs are in `docs/`, [here](https://github.com/pyinvoke/invocations/blob/9e1cae2420c07adc26a0d8dc5b555cbebac7a2f5/invocations/docs.py#L92-L101)
- That default namespace is suitable for drop-in, turnkey reuse elsewhere, such as how Alabaster uses it w/o any doc-related config changes: https://github.com/bitprophet/alabaster/blob/3a4de7b296b782780d71f5a502596d3617a0a69d/tasks.py
- Someone else who keeps their docs in, say, `mydocs` could do the same, but simply add a `'sphinx': {'target': 'mydocs'}` to their `tasks.py` (much as the above alabaster config makes tweaks to other packaging-related subcollections, if you noticed that.)
- My more complex projects have a two-doctree setup, and so `invocations.docs` actually goes even further and does some (currently a bit funky :() stuff with rebinding that default docs collection as two additional Collection objects (`docs` and `www`), which are then imported by e.g. Invoke's own `tasks.py`: https://github.com/pyinvoke/invoke/blob/d1c3df96291b5b16a134a28a948b63db7af6b0f5/tasks.py

**Comment by Aiky30** [comment]
I'll have a good go at running through all of these examples when I next get chance. I guess my confusion was thinking that the configs are all merged / compiled and so would be the same no matter where in the stack the task / NS live.

Thanks for a thorough explanation. 

My confusion was how to package defaults within a packaged app and for the outer tasks (consumers) to see the packaged apps defaults (I have experimented with the seperate app config method). I understand that this is backwards from your use cases and requires the sub apps to know the inner workings of the base app, which as you mention can make debugging difficult and potentially cause other issues. Ref: "At the moment, the intention is for you to distribute your tool as a "packaged" collection or namespace, and for consumers of the tool to include it into their own tasks.py's namespace; then they have the option of overriding your distributed defaults (again in that root namespace object) or not." 

I shall produce a complete example now that I know your intended behaviour. I have spent a lot of time reading the documentation and opened a ticket as last resort / confirmation of your intentions.

Ta
Andrew
